ARCHITECTURE.md

```diff
@@ -0,0 +1,137 @@
+# Architecture overview
+
+This document provides an overview of all parts and pueces of the project as well as how they fit together. It is meant to help new contributors understand where things may be, and how changes in some components may interact with other components of the system.
+
+## Root package (`sqle`)
+
+This is where the engine lives. The engine is the piece that coordinates and makes all other pieces work together as well as the main API users of the system will use to create and configure an engine and perform queries.
+
+Because this is the point where all components fit together, it is also where integration tests are. Those integration tests can be found in `engine_test.go`.
+Every time a feature is added or a bug that comes from a particular query is fixed, a test should be added there.
+
+**How to add integration tests**
+
+The test executing all integration test is `TestQueries`, which you can run with the following command:
+
+```
+go test -run=TestQueries
+```
+
+This test is just executing all the queries in a loop. New test cases should be added to the `queries` package variable at the top of `engine_test.go`.
+Simply add a new element to the slice with the query and the expected result.
+
+## `sql`
+
+This package is probably the most important of the project. It has several main roles:
+- Defines the main interfaces used in the rest of the packages `Node`, `Expression`, ...
+- Provides implementations of components used in the rest of the packages `Row`, `Context`, `ProcessList`, `Catalog`, ...
+- Defines the `information_schema` table, which is a special table available in all databases and contains some data about the schemas of other tables.
+
+### `sql/analyzer`
+
+The analyzer is perhaps the more complex component of the project. It contains a main component, which is the `Analyzer`, in charge of executing its registered rules on execution trees for resolving some parts, removing redundant data, optimizing things for performance, etc.
```

Comment:
Maybe without "perphaps" :) because IMO it's the more complex one by a long shot.